bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/services/seo/LanguageNavigationSiteRootSelectionStrategy.java

```diff
@@ -0,0 +1,146 @@
+        if (languageNavigation != null) {
+            String siteRoot = getPropertyLazy(languageNavigation, LanguageNavigation.PN_NAVIGATION_ROOT, String.class);
+            if (siteRoot != null) {
+                return page.getPageManager().getPage(siteRoot);
+            }
+        }
+        return defaultSelectionStrategy != null ? defaultSelectionStrategy.getSiteRoot(page) : null;
+    }
+
+    @Override
+    public int getStructuralDepth(@NotNull Page page) {
+        Resource languageNavigation = findLanguageNavigation(page.getContentResource());
+        if (languageNavigation != null) {
+            Integer structureDepth = getPropertyLazy(languageNavigation, LanguageNavigation.PN_STRUCTURE_DEPTH, Integer.class);
+            if (structureDepth != null) {
+                return structureDepth;
+            }
+        }
+        return defaultSelectionStrategy != null ? defaultSelectionStrategy.getStructuralDepth(page) : 1;
+    }
+
+    private Resource findLanguageNavigation(Resource contentResource) {
+        Resource templatedResource = contentResource.adaptTo(TemplatedResource.class);
+        if (templatedResource != null) {
+            contentResource = templatedResource;
+        }
+        return findFirst(contentResource,
+            resource -> resource.isResourceType(com.adobe.cq.wcm.core.components.internal.models.v1.LanguageNavigationImpl.RESOURCE_TYPE)
+                || resource.isResourceType(com.adobe.cq.wcm.core.components.internal.models.v2.LanguageNavigationImpl.RESOURCE_TYPE));
+    }
+
+    private Resource findFirst(Resource resource, Predicate<Resource> condition) {
+        if (resource == null) {
+            return null;
+        }
+        // check if the condition is met
+        if (condition.test(resource)) {
+            return resource;
+        }
+        // if not, check for experience fragments and resolve them
+        if (resource.isResourceType(com.adobe.cq.wcm.core.components.internal.models.v1.ExperienceFragmentImpl.RESOURCE_TYPE_V1)
+            || resource.isResourceType(com.adobe.cq.wcm.core.components.internal.models.v1.ExperienceFragmentImpl.RESOURCE_TYPE_V2)) {
+            return Optional.ofNullable(new ExperienceFragmentRequest(resource).adaptTo(ExperienceFragment.class))
+                .map(ExperienceFragment::getLocalizedFragmentVariationPath)
+                .map(xfPath -> resource.getResourceResolver().getResource(xfPath))
+                .map(xfResource -> findFirst(xfResource, condition))
+                .orElse(null);
+        }
+        // if not iterate over the resource's children
+        for (Resource child : resource.getChildren()) {
+            Resource result = findFirst(child, condition);
+            if (result != null) {
+                return result;
+            }
+        }
+
+        return null;
+    }
+
+    private static <T> T getPropertyLazy(Resource resource, String property, Class<T> type) {
```

Comment:
This is probably something that should be in a common helper. Also, the name is a bit misleading, perhaps something more like `getPropertyOrStyle` would be better.

Reply:
Will move it to `Utils` and rename it as suggested.